core/shared/src/main/scala/fs2/concurrent/TimedPull.scala

```diff
@@ -0,0 +1,289 @@
+/*
+ * Copyright (c) 2013 Functional Streams for Scala
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package fs2
+//package concurrent
+
+import fs2.concurrent.SignallingRef
+import scala.concurrent.duration._
+import fs2.internal.Token
+import cats.effect._
+import cats.syntax.all._
+
+object tp {
+  trait TimedPull[F[_], A] {
+    type Timeout
+    def uncons: Pull[F, INothing, Option[(Either[Timeout, Chunk[A]], TimedPull[F, A])]]
+    def startTimer(t: FiniteDuration): Pull[F, INothing, Unit]
+  }
+  object TimedPull {
+    def go[F[_]: Temporal, A, B](pull: TimedPull[F, A] => Pull[F, B, Unit]): Pipe[F, A, B] = { source =>
```

Comment:
Instead of this signature, what do you think of:

```scala
// On ToPull[F[_], O]
def timed[A](pull: TimedPull[F, O] => Pull[F, B, Unit])(implicit t: Temporal[F]): Pull[F, B, Unit]
```

Usage like:
```
s.pull.timed { tp => ... }.stream
```

Reply:
I was thinking of doing both, kinda like `Broadcast` does it. 
I prefer yours for normal usage tbh, my only "issue" with it is that if you go to look at `TimedPull`, there's nothing there showing you how to use it (and you need to refer to `ToPull`)

Reply:
The alternative is to put `TimedPull` inside `Stream`, which is how `StepLeg` does it, and have the signature you propose exclusively. I don't mind that.

Reply:
Yeah I'm going with your suggestion for sure, since it allows reusing things like `repeatPull`